lib/identity_doc_auth/acuant/responses/get_results_response.rb

```diff
@@ -62,6 +65,47 @@ def pii_from_doc
 
         attr_reader :http_response
 
+        def generate_errors
+          return {} if successful_result?
+
+          front_dpi_fail, back_dpi_fail = false
+          front_sharp_fail, back_sharp_fail = false
+          front_glare_fail, back_glare_fail = false
+
+          processed_image_metrics.each do |side, i|
+            hdpi = i['HorizontalResolution'] || 0
+            vdpi = i['VerticalResolution'] || 0
+            if hdpi < DPI_THRESHOLD || vdpi < DPI_THRESHOLD
+              front_dpi_fail = true if side == :front
+              back_dpi_fail = true if side == :back
+            end
+
+            sharpness = i['SharpnessMetric']
+            if sharpness.present? && sharpness < SHARP_THRESHOLD
+              front_sharp_fail = true if side == :front
+              back_sharp_fail = true if side == :back
+            end
+
+            glare = i['GlareMetric']
+            if glare.present? && glare < GLARE_THRESHOLD
+              front_glare_fail = true if side == :front
+              back_glare_fail = true if side == :back
+            end
+          end
+
+          return { results: [Errors::DPI_LOW_BOTH_SIDES] } if front_dpi_fail && back_dpi_fail
+          return { results: [Errors::DPI_LOW_ONE_SIDE] } if front_dpi_fail || back_dpi_fail
```

Comment:
Was there a reason that we have separate "one side failed" vs "both sides failed" errors? I think if we had an array of errors that we added to, or even an ActiveModel::Errors that we add per attribute to, we could get some more useful data?

```suggestion
          errors = Hash.new { |h, k| h[k] = [] }
          
          processed_image_metrics.each do |side, i|
            hdpi = i['HorizontalResolution'] || 0
            vdpi = i['VerticalResolution'] || 0
            if hdpi < DPI_THRESHOLD || vdpi < DPI_THRESHOLD
              errors[side] << :dpi_fail if side == :front
            end

            sharpness = i['SharpnessMetric']
            if sharpness.present? && sharpness < SHARP_THRESHOLD
              errors[side] << :sharp_fail
            end

            glare = i['GlareMetric']
            if glare.present? && glare < GLARE_THRESHOLD
              errors[side] << :glare_fail
            end
          end

          return errors.to_h # or something
```

Reply:
It's a combination of things. The first was that we had slightly different error messages for single errors vs multiple errors, which is in the corresponding IDP PR linked above. The second piece of this was that as part of this I realized that we just needed to refactor the way that we're generating errors for both TrueID and Acuant. My goal for that will be to redesign and simplify the separation between the error generation in the gem and the translation/display in the IDP.

Reply:
A piece I just realized I didn't address: In this case a DPI failure can cause problems with the sharpness and glare metrics. So we aren't delivering multiple errors for that right now. That is also something that will be considered during the refactor.